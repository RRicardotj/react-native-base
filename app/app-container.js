import React from 'react';
import { Provider } from 'react-redux';
import { Navigator } from 'react-native-deprecated-custom-components';

import {
  BackHandler,
  StyleSheet,
  View,
} from 'react-native';

import configureStore from './store/configure-store';
import NavigationBar from './navigation/navigation-bar';
import SceneContainer from './navigation/scene-container';
import Styles from './utils/styles';
import RouteMapper from './navigation/route-mapper';
import Routes from './navigation/routes';

// Store
const store = configureStore();

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Styles.BODY_BG,
  },
  navbar: {
    backgroundColor: Styles.NAVBAR_BG,
  },
});

export default class AppContainer extends React.Component {
  componentWillMount() {
    BackHandler.addEventListener('hardwareBackPress', () => {
      if (this.navigator && this.navigator.getCurrentRoutes().length > 1) {
        this.navigator.pop();
        return true;
      }
      return false;
    });
  }

  renderScene(route, navigator) {
    this.navigator = navigator;

    return (
      <SceneContainer
        title={route.title}
        route={route}
        navigator={navigator}
        onBack={() => {
          if (route.index > 0) {
            navigator.pop();
          }
        }}
        {...this.props}
      />
    );
  }

  render() {
    return (
      <Provider store={store}>
        <View style={styles.container}>
          <Navigator
            initialRoute={Routes.home()}
            renderScene={(route, navigator) => this.renderScene(route, navigator)}
            navigationBar={
              <NavigationBar style={styles.navbar} routeMapper={RouteMapper} />
            }
          />
        </View>
      </Provider>
    );
  }
}
