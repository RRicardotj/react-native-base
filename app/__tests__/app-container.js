import 'react-native';
import renderer from 'react-test-renderer';
import { BackHandler } from 'react-native'; // eslint-disable-line no-duplicate-imports
import React from 'react';
import { shallow } from 'enzyme';

import { AppContainer, mapStateToProps } from '../app-container';

describe('app-container.js', () => {
  const props = {
    dispatch: jest.fn(),
    nav: {
      index: 2,
    },
  };

  beforeEach(() => {
    props.dispatch.mockReset();
  });

  it('test the mapStateToProps function', () => {
    const state = {
      navigation: {},
    };

    const caseProps = mapStateToProps(state);

    expect(caseProps.nav).toEqual({});
  });

  it('renders correctly', () => {
    const tree = renderer.create(<AppContainer {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should press the BackHandler button(with history)', () => {
    let backCallback = null;

    spyOn(
      BackHandler,
      'addEventListener'
    ).and.callFake((callbackName, callback) => {
      backCallback = callback;
    });

    const wrapper = shallow(<AppContainer {...props} />);

    backCallback();

    expect(wrapper).toBeDefined();
    expect(BackHandler.addEventListener).toHaveBeenCalledTimes(1);
    expect(props.dispatch).toHaveBeenCalledTimes(1);
  });

  it('should press the BackHandler button(without history)', () => {
    const caseProps = {
      ...props,
      nav: {
        index: 0,
      },
    };
    let backCallback = null;

    spyOn(
      BackHandler,
      'addEventListener'
    ).and.callFake((callbackName, callback) => {
      backCallback = callback;
    });

    const wrapper = shallow(<AppContainer {...caseProps} />);
    const pop = jest.fn();

    backCallback();

    expect(wrapper).toBeDefined();
    expect(pop).not.toHaveBeenCalled();
    expect(BackHandler.addEventListener).toHaveBeenCalledTimes(1);
    expect(props.dispatch).toHaveBeenCalledTimes(0);

    BackHandler.addEventListener.calls.reset();
  });
});
