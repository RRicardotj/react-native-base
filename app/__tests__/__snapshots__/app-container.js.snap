// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app-container.js renders correctly 1`] = `
<View>
  AppNavigator
</View>
`;
